interfacer/src/browsh/vim_mode.go

```diff
@@ -0,0 +1,594 @@
+			command = vimCommandsBindings[keyCombination]
+		}
+	case insertMode:
+		if isNormalModeKey(ev) {
+			command = "normalMode"
+		}
+	case visualMode:
+		if isNormalModeKey(ev) {
+			command = "normalMode"
+		} else {
+			if ev.Rune() == 'c' {
+				command = "caretMode"
+			}
+			if ev.Rune() == 'o' {
+				//swap cursor position begin->end or end->begin
+			}
+			if ev.Rune() == 'y' {
+				//clipboard
+			}
+		}
+	case caretMode:
+		if isNormalModeKey(ev) {
+			command = "normalMode"
+		} else {
+			switch ev.Key() {
+			case tcell.KeyEnter:
+				generateLeftClick(caretPos.X, caretPos.Y-uiHeight)
+			}
+			switch ev.Rune() {
+			case 'v':
+				command = "visualMode"
+			case 'h':
+				moveVimCaret(func() bool { return caretPos.X > 0 }, &caretPos.X, -1)
+			case 'l':
+				width, _ := screen.Size()
+				moveVimCaret(func() bool { return caretPos.X < width }, &caretPos.X, 1)
+			case 'k':
+				_, height := screen.Size()
+				moveVimCaret(func() bool { return caretPos.Y >= uiHeight }, &caretPos.Y, -1)
+				if caretPos.Y < uiHeight {
+					command = "scrollHalfPageUp"
+					if CurrentTab.frame.yScroll == 0 {
+						caretPos.Y = uiHeight
+					} else {
+						caretPos.Y += (height - uiHeight) / 2
+					}
+				}
+			case 'j':
+				_, height := screen.Size()
+				moveVimCaret(func() bool { return caretPos.Y <= height-uiHeight }, &caretPos.Y, 1)
+				if caretPos.Y > height-uiHeight {
+					command = "scrollHalfPageDown"
+					caretPos.Y -= (height - uiHeight) / 2
+				}
+			}
+		}
+	case makeMarkMode:
+		if unicode.IsLower(ev.Rune()) {
+			if localMarkMap[CurrentTab.ID] == nil {
+				localMarkMap[CurrentTab.ID] = make(map[rune]*mark)
+			}
+			localMarkMap[CurrentTab.ID][ev.Rune()] = makeMark()
+		} else if unicode.IsUpper(ev.Rune()) {
+			globalMarkMap[ev.Rune()] = makeMark()
+		}
+
+		command = "normalMode"
+	case gotoMarkMode:
+		if mark, ok := globalMarkMap[ev.Rune()]; ok {
+			gotoMark(mark)
+		} else if m, ok := localMarkMap[CurrentTab.ID]; unicode.IsLower(ev.Rune()) && ok {
+			if mark, ok := m[ev.Rune()]; ok {
+				gotoMark(mark)
+			}
+		}
+
+		command = "normalMode"
+	case findMode:
+		if isNormalModeKey(ev) {
+			command = "normalMode"
+			findText = ""
+		} else {
+			if ev.Key() == tcell.KeyEnter {
+				currentVimMode = normalMode
+				command = "findText"
+				break
+			}
+			if ev.Key() == tcell.KeyBackspace || ev.Key() == tcell.KeyBackspace2 {
+				if len(findText) > 0 {
+					findText = findText[:len(findText)-1]
+				}
+			} else {
+				findText += string(ev.Rune())
+			}
+		}
+	case linkMode, linkModeNewTab, linkModeCopy:
+		if isNormalModeKey(ev) {
+			command = "normalMode"
+			eraseLinkHints()
+		} else {
+			linkText += string(ev.Rune())
+			updateLinkHintDisplay()
+			if linkModeWithHints {
+				if r, ok := linkHintsToRects[linkText]; ok {
+					if r != nil {
+						switch currentVimMode {
+						case linkMode:
+							if (*r).Height == 2 {
+								generateLeftClickYHack((*r).Left, (*r).Top, true)
```

Comment:
Is this the corner case for the yhack?

Reply:
Yeah. A height of 2 actually implies a height of 1. This is a bit confusing I have to admit, look into parseJSONLinkHints to get some of the logic behind that. Basically we're getting rectangles for the total size of links and we're moving them into a vertical middle position if they're higher than one row, and we're also moving them a bit to the right if there's enough space in that direction. The only positions we need for displaying link hints are r.Top and rLeft, so those are the ones that are being adapted, while r.Height is unchanged.